lib/spack/spack/version.py

```diff
@@ -886,3 +980,170 @@ class VersionError(spack.error.SpackError):
+        if not os.path.exists(dest):
+            tty.die("There was an issue cloning %s to %s" % (self.fetcher.url, dest))
+
+        # Always pull and fetch
+        with working_dir(dest):
+
+            # Don't do anything if already cloned - it might not have remote
+            if not already_cloned:
+                self.fetcher.git("pull")
+                self.fetcher.git("fetch")
+
+            # List tags (refs) by date, so last reference of a tag is newest
+            tags = self.fetcher.git("for-each-ref", "--sort=creatordate", "--format",
+                                    "%(refname)", "refs/tags", output=str).split('\n')
+
+        # Lookup of spack tags to git references
+        tag_lookup = {}
+
+        # Lookup of commits to spack versions
+        commit_to_version = {}
+
+        # For each tag, try to match to a version.
+        versions = [str(v) for v in spack_versions]
+        for tag in tags:
+            tag = tag.replace('refs/tags/', '', 1)
+            for v in versions:
+                if v in tag:
+                    tag_lookup[v] = tag
+
+        # Also prepare to parse versions that aren't in spack
+        for tag in tags:
+            tag = tag.replace('refs/tags/', '', 1)
+
+            # Try to derive a semantic version
+            match = SEMVER_REGEX.match(tag)
+            if match:
+                semver = match.groupdict()['semver']
+                if semver not in tag_lookup:
+                    tag_lookup[semver] = tag
+
+        # Now get commits for these versions
+        with working_dir(dest):
+            for spack_version, tag in tag_lookup.items():
+                command = ["rev-list", "-n", "1", "tags/%s" % tag]
+                commit = self.fetcher.git(*command, output=str)
+                commit_to_version[commit.strip()] = spack_version
+
+            # Get list of all commits, this is in reverse order
+            commits = self.fetcher.git("log", "--all", "--pretty=format:%H", output=str)
+
+        # commits[0] is most recent, commits[-2] is first
+        commits = [c for c in commits.split('\n') if c]
+
+        for commit in commits:
+            # Get commits associated with tags that're ancestors of this commit
+            # Values are tuples of ancestor and distance in commits
+            ancestor_commits = []
+            for tag_commit in commit_to_version:
+                try:
+                    with working_dir(dest):
+                        self.fetcher.git(
+                            'merge-base', '--is-ancestor', tag_commit, commit)
+                        distance = self.fetcher.git(
+                            'rev-list', '%s..%s' % (tag_commit, commit), '--count',
+                            output=str, error=str).strip()
+                    ancestor_commits.append((tag_commit, int(distance)))
+                except spack.util.executable.ProcessError:
+                    # is-ancestor check will raise a ProcessError when False
```

Comment:
would probably clearer to use fail_on_error=False for the call to merge-base, followed by a conditional

Reply:
Used `ignore_errors` instead so we still raise on unexpected error codes.